src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Interfaces/IVsTestConsoleWrapper.cs

```diff
@@ -19,112 +20,307 @@ public interface IVsTestConsoleWrapper : IVsTestConsoleWrapperAsync
+        /// <param name="eventsHandler">The session event handler.</param>
+        /// 
+        /// <returns>A test session info object.</returns>
+        ITestSession StartTestSession(
+            IList<string> sources,
+            string runSettings,
+            ITestSessionEventsHandler eventsHandler);
 
         /// <summary>
-        /// Starts a test run given a list of sources by giving caller an option to start their own test host.
+        /// Starts a new test session.
         /// </summary>
-        /// <param name="sources">Sources to Run tests on</param>
-        /// <param name="runSettings">RunSettings XML to run the tests</param>
-        /// <param name="options">Options to be passed into the platform.</param>
-        /// <param name="testRunEventsHandler">EventHandler to receive test run events</param>
-        /// <param name="customTestHostLauncher">Custom test host launcher for the run.</param>
-        void RunTestsWithCustomTestHost(IEnumerable<string> sources, string runSettings, TestPlatformOptions options, ITestRunEventsHandler testRunEventsHandler, ITestHostLauncher customTestHostLauncher);
+        /// 
+        /// <param name="sources">The list of source assemblies for the test run.</param>
+        /// <param name="runSettings">The run settings for the run.</param>
+        /// <param name="options">The test platform options.</param>
+        /// <param name="eventsHandler">The session event handler.</param>
+        /// 
+        /// <returns>A test session info object.</returns>
+        ITestSession StartTestSession(
+            IList<string> sources,
+            string runSettings,
+            TestPlatformOptions options,
+            ITestSessionEventsHandler eventsHandler);
 
         /// <summary>
-        /// Starts a test run given a list of test cases by giving caller an option to start their own test host
+        /// Starts a new test session.
         /// </summary>
-        /// <param name="testCases">TestCases to run.</param>
-        /// <param name="runSettings">RunSettings XML to run the tests.</param>
-        /// <param name="testRunEventsHandler">EventHandler to receive test run events.</param>
-        /// <param name="customTestHostLauncher">Custom test host launcher for the run.</param>
-        void RunTestsWithCustomTestHost(IEnumerable<TestCase> testCases, string runSettings, ITestRunEventsHandler testRunEventsHandler, ITestHostLauncher customTestHostLauncher);
+        /// 
+        /// <param name="sources">The list of source assemblies for the test run.</param>
+        /// <param name="runSettings">The run settings for the run.</param>
+        /// <param name="options">The test platform options.</param>
+        /// <param name="eventsHandler">The session event handler.</param>
+        /// <param name="testHostLauncher">The custom host launcher.</param>
+        /// 
+        /// <returns>A test session info object.</returns>
+        ITestSession StartTestSession(
```

Comment:
Restarting sessions shouldn't be a thing.

The current PR enables spawning a session that has one or more testhosts attached to it and only one request per session can be served. This is because the default behavior is to terminate the testhost(s) once the request (discovery/run) was successfully completed.

The next round of changes should enable testhost sharing between discovery and run and should disable the default testhost termination behavior. This means the testhost(s) will be kept alive until the session is terminated with an explicit call from the API user (TW for example).

The only possible need for session restarting would be when the runsettings/list of sources change between discovery and execution. This will require restarting the testhosts, but at this point I think we're better off with terminating the current session and just starting a new session with the updated settings/list of sources. From the performance standpoint the cost of those operations should be about the same.

Reply:
I meant stopping a session and creating a new one.

besides the settings/list of sources would changing between debugging and profiling also require us to create a new session?

Reply:
@drognanar Debugging doesn't use the old workflow anymore (i.e. `vstest.console` asks `VS` to launch the `testhost` process and attach to it) unless we're in compatibility mode. Instead the new workflow launches the `testhost` as a standalone process and only asks `VS` to attach to it just before starting the test execution, provided `IsDebug` flag is set of course. No debugging takes place for test discovery though.

For profiling though we still use the old workflow outlined above. As such we ask `VS` to launch the `testhost` process on our behalf because there's some specific setup that needs to take place first (setting env vars, etc.). Now this behavior is achievable by passing along an object implementing `ITestHostLauncher` to the `StartTestSession` call, the downside being profiling will be set up for test discovery too in addition to test execution. If we wish to avoid this behavior though, the only way to do that would be to start the test session with no `ITestHostLauncher` (in this way test discovery won't be profiled) and initiate test execution by re-launching the `testhost` process in profiling mode.